### PR TITLE
[improve][broker] Optimize message payload traffic for ShadowReplicator

### PR DESCRIPTION
**Pull request by RobertIndie**
## Motivation

Currently, the ShadowReplicator syncs all message payloads from the source topic to the shadow topic (ShadowManagedLedger). This significantly increases traffic between brokers. However, the ShadowManagedLedger only requires the position information from these messages. It can retrieve the payload directly from the Bookkeeper. Therefore, the ShadowReplicator does not need to sync the message payloads between brokers.

### Modifications

- Prevent ShadowReplicator from syncing message payloads

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/RobertIndie/pulsar/pull/16

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/ShadowReplicatorTest.java [hunk 1]
```diff
@@ -131,7 +131,7 @@ public void testShadowReplication() throws Exception {
 
         Message<byte[]> shadowMessage = shadowConsumer.receive(5, TimeUnit.SECONDS);
 
-        Assert.assertEquals(shadowMessage.getData(), sourceMessage.getData());
+        Assert.assertEquals(shadowMessage.getData().length, 0);
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/ShadowReplicatorTest.java:134]: This is a breaking change. The shadow consumer should still receive the message

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/ShadowTopicRealBkTest.java [hunk 2]
```diff
@@ -74,7 +75,7 @@ public void cleanup() throws Exception {
 
     @Test
     public void testReadFromStorage() throws Exception {
-        final var sourceTopic = TopicName.get("test-read-from-source").toString();
+        final var sourceTopic = TopicName.get("test-read-from-source" + UUID.randomUUID()).toString();
```
[the remark this answers is not in the document]
Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/ShadowTopicRealBkTest.java:78]: This is to fix the flaky test failure here: https://github.com/RobertIndie/pulsar/actions/runs/10626045164/job/29458947386#step:11:1369

The test appears to run twice, leading to an error that the topic already exists.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java [hunk 3]
```diff
@@ -239,7 +239,8 @@ public void run() {
         ManagedLedgerImpl.TOTAL_SIZE_UPDATER.addAndGet(ml, dataLength);
 
         long ledgerId = ledger != null ? ledger.getId() : ((Position) ctx).getLedgerId();
-        if (ml.hasActiveCursors()) {
+        // Don't insert to the entry cache for the ShadowManagedLedger
+        if (!(ml instanceof ShadowManagedLedgerImpl) && ml.hasActiveCursors()) {
```
[the remark this answers is not in the document]
Comment (RobertIndie) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java]: Good suggestion. Applied it.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/ShadowReplicatorTest.java (hunk 1 above)
Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/ShadowReplicatorTest.java]: Sorry. I changed it by mistake. The test should fail. However, I found that this test class has never been tested by the CI. Will fix it.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (RobertIndie) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:243]: As we discussed privately, this change will break the behavior. The message with the empty payload can't be inserted into the entry cache even without the ShadowReplicator. Some tests also failed.
So we decided to revert to the original `instanceof` solution.

**Review by BewareMyPower — APPROVED** [review]

**Comment by lhotari** [comment]
It looks like this would require #22972 changes for applying this cleanly to branch-3.2 .

**Pull request merged**